### PR TITLE
feat(migration): rewrite secret backend UUIDs during model import

### PR DESCRIPTION
**Pull request by nvinuesa**
When a model migrates to a new controller, each secret backend on the target is given its own random 
identifier and is matched to the source only by name. The new-format import copies the model's secret 
rows as they were on the source, so the backend identifiers in those rows still point at the source 
controller and mean nothing on the target. Left alone, every external secret would reference a backend
that does not exist here.

This change fixes those identifiers before the model data is written. Once the controller-level data 
has been imported, a small step looks up each backend by name on the target, builds the mapping from 
each secret revision to the target's backend identifier, and rewrites the payload to use it, right 
before the model database insert. If a revision has no matching backend the import stops with a clear 
error, and because this runs before anything is written to the model database, nothing is left half 
applied. It sits at the import stage, rather than at activation like the cross-model relation work, 
because the target backends already exist and can be resolved by name at this point.

The behavior is covered by unit tests for the rewrite itself and by real-database tests that import a
model with an external backend and confirm the stored rows end up with the target's identifier, along
with a negative case proving that an unmapped revision fails without writing anything. Nothing else 
on the import path changes, so this sets up the later activation and secret re-attach work to rely on 
correct target identifiers.

## QA steps

Not fully wired yet, unit tests must pass though.

## Links

**Jira card:** [JUJU-9908](https://warthogs.atlassian.net/browse/JUJU-9908)

[JUJU-9908]: https://warthogs.atlassian.net/browse/JUJU-9908?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/migration/secretrewrite.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// Import returns a generic import failure → the source enters ABORT → Task 11
+// AbortImport / RemoveOnAbortImport cleans the controller-DB data and the
+// partial model. No model-DB rows were written, so the rewrite needs no
+// compensation of its own.
+func RewriteSecretBackendUUIDs(payload *latest.ModelExport, revisionToTargetBackend map[string]string) error {
+	if payload == nil {
+		return nil
+	}
+
+	if payload.SecretValueRef == nil && payload.SecretDeletedValueRef == nil {
```
Comment (gfouillet) [inline internal/migration/secretrewrite.go]: **todo:** The emptiness check uses `== nil` rather than the idiomatic `len(s) == 0`. Non-nil but empty slices would skip the early return and enter the loops (which iterate zero times), so the code is correct either way. Consider `len(payload.SecretValueRef) == 0 && len(payload.SecretDeletedValueRef) == 0` to catch both nil and empty.

File: internal/migration/secretrewrite_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+// secretRewriteSuite exercises [migration.RewriteSecretBackendUUIDs].
+type secretRewriteSuite struct {
+	testhelpers.IsolationSuite
+}
+
+func TestSecretRewriteSuite(t *testing.T) {
+	tc.Run(t, &secretRewriteSuite{})
+}
+
+func (s *secretRewriteSuite) TestRewriteSecretBackendUUIDs_NilPayload(c *tc.C) {
```
Comment (gfouillet) [inline internal/migration/secretrewrite_test.go]: **todo:** Test function names use underscores (`TestRewriteSecretBackendUUIDs_NilPayload`), while the sibling `modelimporter_test.go` in the same package uses standard CamelCase (`TestImportModelSecretBackendRewrite`). Consider using CamelCase or `t.Run("nil payload", ...)` subtests for consistency.

File: internal/migration/modelimporter_test.go [hunk 3]
```diff
@@ -109,3 +112,192 @@ func (s *modelImporterSuite) TestImportModel(c *tc.C) {
...
+	// Source backend UUID — different from the target's.
+	sourceBackendUUID := uuid.MustNewUUID().String()
+
+	now := time.Now().UTC()
+	secretID := uuid.MustNewUUID().String()
+	revUUID1 := uuid.MustNewUUID().String()
+	revUUID2 := uuid.MustNewUUID().String()
+	passwordHash := "some-hash"
+
+	payload := &v4_1_0.ModelExport{
```
Comment (gfouillet) [inline internal/migration/modelimporter_test.go]: **suggestion:** The payload is constructed as `&v4_1_0.ModelExport{...}` instead of `&latest.ModelExport{...}`. The production code types the payload as `*latest.ModelExport`, and `domain/export/types/latest/doc.go` recommends using `latest` for code operating on the target payload. The sibling `secretrewrite_test.go` correctly uses `&latest.ModelExport{}`. Since `latest.ModelExport` is a type alias for `v4_1_0.ModelExport`, this compiles fine, but a future payload-version bump would force unnecessary edits here.

File: internal/migration/secretrewrite_test.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+	// Value refs rewritten.
+	c.Check(payload.SecretValueRef[0].BackendUUID, tc.Equals, "target-1")
+	c.Check(payload.SecretValueRef[0].RevisionID, tc.Equals, "rid-1") // unchanged
+	c.Check(payload.SecretValueRef[1].BackendUUID, tc.Equals, "target-2")
+	c.Check(payload.SecretValueRef[1].RevisionID, tc.Equals, "rid-2") // unchanged
+
+	// Deleted value refs rewritten.
+	c.Check(payload.SecretDeletedValueRef[0].BackendUUID, tc.Equals, "target-1")
+	c.Check(payload.SecretDeletedValueRef[0].RevisionID, tc.Equals, "rid-1") // unchanged
```
Comment (gfouillet) [inline internal/migration/secretrewrite_test.go]: **todo:** Several comment lines in this file exceed the 80-character limit (e.g. line 76: `c.Check(payload.SecretDeletedValueRef[0].RevisionID, tc.Equals, "rid-1") // unchanged`). The trailing `// unchanged` comments are also somewhat redundant given the preceding section comment. Worth wrapping or trimming before merge.

File: domain/secretbackend/service/import.go [hunk 5]
```diff
@@ -31,23 +31,62 @@ func (s *Service) ImportSecretBackendReferences(
...
 			backendID = backend.ID
 			backendIDs[ref.BackendName] = backendID
 		}
-
-		if _, err := s.st.AddSecretBackendReference(
-			ctx, &secrets.ValueRef{BackendID: backendID}, modelUUID, ref.SecretRevisionUUID, ref.SecretID,
-		); err != nil {
-			return errors.Errorf("adding secret backend reference for revision %q: %w", ref.SecretRevisionUUID, err)
-		}
+		revisionMap[ref.SecretRevisionUUID] = backendID
```
Comment (gfouillet) [inline domain/secretbackend/service/import.go:100]: **thought:** In `resolveBackendUUIDsByRevision`, if two refs share the same `SecretRevisionUUID` but reference different `BackendName`, the second would silently overwrite the first in `revisionMap` (line 89: `revisionMap[ref.SecretRevisionUUID] = backendID`). This is unlikely in practice (a revision belongs to one backend), but the behavior is untested. A test documenting the intended behavior (last-wins or reject-as-invalid) would make the contract explicit.

File: internal/migration/modelimporter_test.go [hunk 6]
```diff
@@ -109,3 +112,192 @@ func (s *modelImporterSuite) TestImportModel(c *tc.C) {
...
+	c.Assert(err, tc.ErrorIsNil)
+	c.Check(insertedBackendUUID, tc.Equals, targetBackendUUID,
+		tc.Commentf("secret_deleted_value_ref should carry target backend UUID, got %q", insertedBackendUUID))
+	c.Check(insertedRevisionID, tc.Equals, "ext-rev-2")
+}
+
+// TestImportModelSecretBackendRewriteMissingRef verifies that a
+// secret_value_ref revision absent from SecretBackendRefs causes
+// ImportModel to error before committing any model-DB rows.
+func (s *modelImporterSuite) TestImportModelSecretBackendRewriteMissingRef(c *tc.C) {
```
Comment (gfouillet) [inline internal/migration/modelimporter_test.go:285]: **todo:** `TestImportModelSecretBackendRewriteMissingRef` omits `SecretBackendRefs` entirely (hitting the "nil map" branch). It would be good to add a test where `SecretBackendRefs` is provided but references a backend name that doesn't exist on the target — a distinct error path through `GetSecretBackend` -> NotFound. Assert the error matches `resolving target secret backends.*looking up secret backend "nonexistent"` and zero `secret_value_ref` rows.

File: internal/migration/modelimporter_test.go [hunk 7]
```diff
@@ -109,3 +112,192 @@ func (s *modelImporterSuite) TestImportModel(c *tc.C) {
...
+		return err
+	})
+	c.Assert(err, tc.ErrorIsNil)
+	return backendUUID
+}
+
+// TestImportModelSecretBackendRewrite verifies that secret_value_ref and
+// secret_deleted_value_ref backend_uuid fields are rewritten from the source
+// controller's backend UUIDs to the target's before the model-DB insert.
+func (s *modelImporterSuite) TestImportModelSecretBackendRewrite(c *tc.C) {
```
Comment (gfouillet) [inline internal/migration/modelimporter_test.go:187]: **todo:** `TestImportModel` uses a nil `ModelDBPayload` (returns before `reconcileSecretBackendUUIDs` is called). It would be good to add a test exercising a non-nil payload with model data but no `SecretValueRef`/`SecretDeletedValueRef` — the common case of a model with no external-backed secrets — to verify the reconcile + insert path when there's nothing to rewrite.

File: domain/secretbackend/service/import.go [hunk 8]
```diff
@@ -31,23 +31,62 @@ func (s *Service) ImportSecretBackendReferences(
 		return nil
 	}
 
+	revisionMap, err := s.resolveBackendUUIDsByRevision(ctx, refs)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	for _, ref := range refs {
+		if _, err := s.st.AddSecretBackendReference(
```
Comment (gfouillet) [inline domain/secretbackend/service/import.go:40]: **suggestion:** I think we can get ride of `secrets.ValueRef` embbeding for the backendID, other fields are never used. Just pass a pointer to the backendID (or maybe we don't even need a pointer)

**Review by gfouillet — COMMENTED** [review]
Mostly LGTM, I have few comments

File: internal/migration/modelimporter_test.go (hunk 3 above)
Comment (nvinuesa) [reply internal/migration/modelimporter_test.go]: I rebased but forgot to fix the newly written tests. Fixing now.

**Review by gfouillet — APPROVED** [review]
Thanks

**Review by iyiguncevik — APPROVED** [review]

File: internal/migration/secretrewrite.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+// A value-ref / deleted-value-ref revision with no mapping is a hard error
+// (the model-DB insert has not run yet, so no rows leak).
+//
+// The rewrite runs after the controller-data import and before any model-DB
+// write. If it errors (missing mapping), the caller returns the error. The v8
+// import returns a generic import failure, the source enters ABORT, and Task
+// 11 AbortImport / RemoveOnAbortImport cleans the controller-DB data and the
+// partial model. No model-DB rows were written, so the rewrite needs no
+// compensation of its own.
+func RewriteSecretBackendUUIDs(payload *latest.ModelExport, revisionToTargetBackend map[string]string) error {
```
Comment (SimonRichardson) [inline internal/migration/secretrewrite.go:58]: Why is this public?

**Review by SimonRichardson — COMMENTED** [review]
## Findings

### 1. [critical] `secret_deleted_value_ref` rows will always be unmapped, blocking migration

**Why it is a problem.** `RewriteSecretBackendUUIDs` hard-errors on any
`secret_deleted_value_ref` revision that has no entry in
`revisionToTargetBackend`. The map is built from
`info.SecretBackendRefs`, which is sourced from the controller-DB
`secret_backend_reference` table. But the source controller deletes the
`secret_backend_reference` row for a revision at the same time it moves the
`secret_value_ref` row to `secret_deleted_value_ref`. Therefore, any
`secret_deleted_value_ref` row present at migration time is guaranteed to have
no corresponding `SecretBackendRef` entry, and the import will hard-error.

This is normal steady state, not a transient window: if the external backend
(e.g., Vault) is unreachable during a prune cycle, `secret_deleted_value_ref`
rows persist indefinitely with no `secret_backend_reference` counterpart. There
is no retry mechanism, because `secret_revision` is also deleted, so the
orphaned `secret_deleted_value_ref` rows are unreachable through the normal
secret model.

**Where it appears.**

- `internal/migration/secretrewrite.go:62-67` — the `SecretDeletedValueRef`
  loop that hard-errors:
  ```go
  for i := range payload.SecretDeletedValueRef {
      rev := payload.SecretDeletedValueRef[i].RevisionUUID
      targetBackend, ok := revisionToTargetBackend[rev]
      if !ok {
          return errors.Errorf(
              "no target secret backend for secret revision %q (deleted value ref)", rev)
      }
      payload.SecretDeletedValueRef[i].BackendUUID = targetBackend
  }
- `domain/removal/service/secret.go:238-257` (deleteSecretRevisions) — the
deletion flow that moves secret_value_ref → secret_deleted_value_ref in
the model DB, then unconditionally calls
controllerState.RemoveSecretBackendReference(ctx, deletedRevisionUUIDs...)
(line 251), removing the secret_backend_reference row.
- domain/removal/service/secret.go:260-289 (deleteSecretRevisionContent) —
best-effort external content deletion; logs warnings on failure and returns
nil, so RemoveSecretBackendReference always runs afterward.
- domain/modelmigration/state/controller/state.go:1409-1419
(getSecretBackendRefs) — reads secret_backend_reference; does not include
deleted revisions.
- domain/export/state/model/export.go:772 — exports ALL
secret_deleted_value_ref rows unconditionally.
Evidence — the deletion lifecycle.
```
deleteSecretRevisions (service)                         [secret.go:238]
├── modelState.DeleteSecretRevisions (model DB)         [secret.go:240]
│   ├── INSERT OR IGNORE INTO secret_deleted_value_ref  [secret.go:452-455]
│   │   SELECT revision_uuid, backend_uuid, revision_id
│   │   FROM secret_value_ref WHERE revision_uuid IN (...)
│   ├── DELETE FROM secret_value_ref WHERE ...          [secret.go:456]
│   └── DELETE FROM secret_revision WHERE ...           [secret.go:458]
├── deleteSecretRevisionContent (best-effort!)          [secret.go:246]
│   ├── sb.DeleteContent(ctx, ref)        ← logs warning on failure, continues
│   └── modelState.DeleteUserSecretRevisionRef(ctx, ref) ← logs warning, continues
└── controllerState.RemoveSecretBackendReference(ctx, deletedRevisionUUIDs...)  [secret.go:251]
    └── DELETE FROM secret_backend_reference WHERE secret_revision_uuid IN (...)
```
After the job completes:
```
Table
secret_deleted_value_ref
secret_value_ref
secret_revision
secret_backend_reference
```
Evidence — the information gap is fundamental.
The secret_deleted_value_ref export type carries only the source backend UUID,
not the backend name:
```
// domain/export/types/v4_1_0/model.go:1221
type SecretDeletedValueRef struct {
    RevisionUUID string `db:"revision_uuid"`
    BackendUUID  string `db:"backend_uuid"`   // source controller's UUID
    RevisionID   string `db:"revision_id"`
}
```
The target cannot resolve the source BackendUUID to a name because:
1. The secret_backend_reference row (the only controller-DB link between
revision UUID and backend name) has been deleted.
2. The target doesn't have the source controller's secret_backend table.
3. The source's secret_backend table (uuid → name) is in the controller DB,
but secret_deleted_value_ref is in the model DB — the source-side export
code can't join across databases.
The information needed to rewrite secret_deleted_value_ref rows simply does
not exist in the migration envelope.
Evidence — the PR's own test is unrealistic.
```
TestImportModelSecretBackendRewrite (modelimporter_test.go:245) puts
revUUID2 in both SecretDeletedValueRef and SecretBackendRefs:
SecretBackendRefs: []coremodelmigration.SecretBackendReference{
    {BackendName: targetBackendName, SecretRevisionUUID: revUUID1, SecretID: secretID},
    {BackendName: targetBackendName, SecretRevisionUUID: revUUID2, SecretID: secretID},
},
```
This pairing cannot occur in production. On the source controller, when a
revision moves to secret_deleted_value_ref, its secret_backend_reference
row is removed. A revision is never in both tables simultaneously. The test
validates an impossible state.
Impact. Any model using an external secret backend (Vault, k8s secrets)
that has gone through at least one secret revision deletion cycle where
external content deletion transiently failed will fail migration with a hard
error. This is not a rare edge case — it is the expected state for any model
with auto-prune enabled (the default for user secrets) and a temporarily
unavailable external backend. The controller-DB import has already run, so
RemoveOnAbortImport must clean it up. No model-DB rows are written (the error
precedes the insert), but the migration is blocked.
